Added modification to the teleportation to include tamed entities

KrisTheCanadian opened the pull request:
This pull request includes several updates to the `gradle.properties` file and enhancements to the teleportation functionality in the `EssentialCommands` mod. The most important changes include version updates for dependencies, new configuration options for teleporting with followers, and the implementation of the teleportation logic for followers.

Updates to dependencies:

* [`gradle.properties`](diffhunk://#diff-3d103fc7c312a3e136f88e81cef592424b8af2464c468116545c4d22d6edcf19L6-R11): Updated `minecraft_version`, `yarn_mappings`, and `fabric_version` to 1.21.4 versions.
* [`gradle.properties`](diffhunk://#diff-3d103fc7c312a3e136f88e81cef592424b8af2464c468116545c4d22d6edcf19L35-R35): Updated `mr_game_versions` and `cf_game_versions` to 1.21.4. [[1]](diffhunk://#diff-3d103fc7c312a3e136f88e81cef592424b8af2464c468116545c4d22d6edcf19L35-R35) [[2]](diffhunk://#diff-3d103fc7c312a3e136f88e81cef592424b8af2464c468116545c4d22d6edcf19L44-R44)

Enhancements to teleportation functionality:

* [`EssentialCommandsConfig.java`](diffhunk://#diff-8a73c6cf40f7f1148552f26ca70aca9253b2a4436e5dbdc4cbc7959a90640443R85-R86): Added new configuration options `ALLOW_TELEPORT_WITH_FOLLOWERS` and `TELEPORT_WITH_FOLLOWERS_RADIUS`.
* [`EssentialCommandsConfigSnapshot.java`](diffhunk://#diff-8e88362fc306066fd039f70abc82d911b49295b66c8c6bcbe10b7e17a7e66c1fR62-R63): Updated snapshot class to include new teleportation options. [[1]](diffhunk://#diff-8e88362fc306066fd039f70abc82d911b49295b66c8c6bcbe10b7e17a7e66c1fR62-R63) [[2]](diffhunk://#diff-8e88362fc306066fd039f70abc82d911b49295b66c8c6bcbe10b7e17a7e66c1fR138-R139)
* [`PlayerTeleporter.java`](diffhunk://#diff-429befddb081f2f83fb26d11a0b9761e532a18fe1cb5c0ba8402fe1cdbec1aaeL8-R21): Implemented logic to teleport tamed followers along with the player, including checks for follower radius and tracking teleported followers. [[1]](diffhunk://#diff-429befddb081f2f83fb26d11a0b9761e532a18fe1cb5c0ba8402fe1cdbec1aaeL8-R21) [[2]](diffhunk://#diff-429befddb081f2f83fb26d11a0b9761e532a18fe1cb5c0ba8402fe1cdbec1aaeL64-R128)

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/John-Paul-R/Essential-Commands/313)
<!-- Reviewable:end -->

KrisTheCanadian committed:
- Teleportation now includes followers by default

KrisTheCanadian committed:
chunk not updated fix

KrisTheCanadian commented:
WIP still working on teleporting pets between worlds

John-Paul-R commented:
Very cool stuff! Definitely interested in merging when this is done. Thanks for building this!

If you're hitting a technical hurdle, feel free to send a message in the discord, happy to brainstorm.

KrisTheCanadian committed:
- added a way to teleport tamedEntities to different dimensions

- refactored code + added javadocs

KrisTheCanadian commented:
Should be ready for review. I was able to figure out a way to save the tameable entity data and recreate them in the other dimension. I've tested it on my end and everything seems good.

John-Paul-R reviewed (changes requested):
In general, looks good to me. I'll test this out myself shortly. Thanks for putting this together!

> *[Reviewable](https://reviewable.io/reviews/John-Paul-R/Essential-Commands/313#-:-OIrhaaz89EzpbpYL6o4:b-6r54ef)* status: 0 of 4 files reviewed, 4 unresolved discussions

___
*[`src/main/java/com/fibermc/essentialcommands/teleportation/PlayerTeleporter.java` line 95 at r2](https://reviewable.io/reviews/John-Paul-R/Essential-Commands/313#-OIrf14Y1xSL7S2r-ogH:-OIrf14Y1xSL7S2r-ogI:b-z3xrux) ([raw file](https://github.com/John-Paul-R/Essential-Commands/blob/7eb67fb665ef8c5a0eb98bd31711df109dddc28a/src/main/java/com/fibermc/essentialcommands/teleportation/PlayerTeleporter.java#L95)):*
> ```Java
>         Vec3d targetVec = new Vec3d(dest.pos().x, dest.pos().y, dest.pos().z);
> 
>         List<TameableEntity> pets = detectTamedPets(playerEntity, playerPos);
> ```

I think I'd expect that we don't load any entities if `TELEPORT_FOLLOWERS` is false, but I don't see that configuration option handled right now.
___
*[`src/main/java/com/fibermc/essentialcommands/teleportation/PlayerTeleporter.java` line 122 at r2](https://reviewable.io/reviews/John-Paul-R/Essential-Commands/313#-OIrg7sD40v_qzOLdfVV:-OIrg7sD40v_qzOLdfVW:bw9atl9) ([raw file](https://github.com/John-Paul-R/Essential-Commands/blob/7eb67fb665ef8c5a0eb98bd31711df109dddc28a/src/main/java/com/fibermc/essentialcommands/teleportation/PlayerTeleporter.java#L122)):*
> ```Java
> 
>             LOGGER.info("Checking pet {} ({}) - Tamed: {}, Owner Matches: {}, Sitting: {}",
>                 pet.getType().getTranslationKey(), pet.getUuid(), isTamed, isSameOwner, isSitting);
> ```

While probably useful in development, I feel this may end up unnecessarily noisy for users.
___
*[`src/main/java/com/fibermc/essentialcommands/teleportation/PlayerTeleporter.java` line 175 at r2](https://reviewable.io/reviews/John-Paul-R/Essential-Commands/313#-OIrgcum3AUF4oJTnPcl:-OIrgcum3AUF4oJTnPcm:bfdf1vo) ([raw file](https://github.com/John-Paul-R/Essential-Commands/blob/7eb67fb665ef8c5a0eb98bd31711df109dddc28a/src/main/java/com/fibermc/essentialcommands/teleportation/PlayerTeleporter.java#L175)):*
> ```Java
>             return true;
>         } else {
>             LOGGER.error("Failed to create entity from NBT for pet ({})!", pet.getUuid());
> ```

What are the possible failure modes here? Is is possible that an entity gets created, but isn't a `TameableEntity`? If so, ought we need to handle that by snapping that new one out of existence?

Entirely possible this is a non issue, or at least effectively impossible or practical purposes. I'll defer to your judgement!
___
*[`src/main/java/com/fibermc/essentialcommands/config/EssentialCommandsConfig.java` line 85 at r2](https://reviewable.io/reviews/John-Paul-R/Essential-Commands/313#-OIrfeubDpsihbfOqrrG:-OIrfeuc-TTB9nevHxla:b-v1cqtw) ([raw file](https://github.com/John-Paul-R/Essential-Commands/blob/7eb67fb665ef8c5a0eb98bd31711df109dddc28a/src/main/java/com/fibermc/essentialcommands/config/EssentialCommandsConfig.java#L85)):*
> ```Java
>     @ConfigOption public final Option<Double>  TELEPORT_INTERRUPT_ON_MOVE_AMOUNT = new Option<>("teleport_interrupt_on_move_max_blocks", 3D, ConfigUtil::parseDouble);
>     @ConfigOption public final Option<Boolean> ALLOW_TELEPORT_BETWEEN_DIMENSIONS = new Option<>("allow_teleport_between_dimensions", true, Boolean::parseBoolean);
>     @ConfigOption public final Option<Boolean> ALLOW_TELEPORT_WITH_FOLLOWERS = new Option<>("allow_teleport_with_followers", true, Boolean::parseBoolean);
> ```

I generally try to avoid making changes that would cause existing features to change how they behave without user intervention, so I'd prefer to default this to false for now.



_Suggestion:_
```Java
false,
```


<!-- Sent from Reviewable.io -->

John-Paul-R reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/John-Paul-R/Essential-Commands/313)* status: 0 of 4 files reviewed, 5 unresolved discussions (waiting on @KrisTheCanadian)

___
*[`src/main/java/com/fibermc/essentialcommands/config/EssentialCommandsConfigSnapshot.java` line 138 at r2](https://reviewable.io/reviews/John-Paul-R/Essential-Commands/313#-OIrk6-L6tKoa583ATcl:-OIrk6-L6tKoa583ATcm:b-gbigz5) ([raw file](https://github.com/John-Paul-R/Essential-Commands/blob/7eb67fb665ef8c5a0eb98bd31711df109dddc28a/src/main/java/com/fibermc/essentialcommands/config/EssentialCommandsConfigSnapshot.java#L138)):*
> ```Java
>         this.TELEPORT_INTERRUPT_ON_MOVE_AMOUNT  = config.TELEPORT_INTERRUPT_ON_MOVE_AMOUNT.getValue();
>         this.ALLOW_TELEPORT_BETWEEN_DIMENSIONS  = config.ALLOW_TELEPORT_BETWEEN_DIMENSIONS.getValue();
>         this.TELEPORT_FOLLOWERS                 = config.ALLOW_TELEPORT_WITH_FOLLOWERS.getValue();
> ```

As currently written, I think I'd prefer this to be named simply `TELEPORT_WITH_FOLLOWERS`, since I believe this'll affect how every teleport occurs, it doesn't just allow an optional-per-teleport thing.



_Suggestion:_
```Java
TELEPORT_WITH_FOLLOWERS
```


<!-- Sent from Reviewable.io -->

KrisTheCanadian committed:
Added sanity check incase NBT fails to create the new one

- added the forgotten feature flag TELEPORT_FOLLOWERS

KrisTheCanadian commented:
@John-Paul-R 

- I've added the missing feature flag TELEPORT_FOLLOWERS which was accidentally removed while testing. 
- I removed the logger info as requested (forgot to remove it). 
- Changed the default setting of TELEPORT_FOLLOWERS to false as requested
- Changed the feature name from ALLOW_TELEPORT_WITH_FOLLOWERS to TELEPORT_WITH_FOLLOWERS

In regards to the issue where the entity gets created but isn't a Tameable entity, this should be a non-issue since we do look for only entities that are tameable that are sitting and which have a owner. The only thing that could occur is that the transfer between dimensions fails for some reason and it's not able to copy to old entity and results cancelling the creation of the new pet. I went ahead and added a sanity check to make sure the newTamedPet (pet in the new dimension) is created before removing the old one to avoid data loss. 

I did not re-test the logic as it's mostly the same. Let me know if everything works on your side when you're testing. Also, i usually tend to write tests for this sort of thing but i'm a bit new to minecraft modding in general so i'll have to check out how i would go about writing those.

John-Paul-R commented:
Alright, I've reviewed and verified the changes work. This looks good to me!

Thanks you again for putting this together. This PR was very professionally done. Looks clean!

John-Paul-R commented:
Also, thanks for updating the MC publishing versions!
